fieldservice_sale/tests/test_fsm_sale_order.py

```diff
@@ -0,0 +1,418 @@
+        )
+        # Sale Order linked to FSM order
+        self.assertEqual(
+            len(fsm_order.ids), 1, "FSM Sale: Sale Order not linked to FSM Order"
+        )
+
+        # Complete the FSM order
+        if self._isp_account_installed():
+            fsm_order = self._fulfill_order(fsm_order)
+        fsm_order.write(
+            {"date_end": fields.Datetime.today(), "resolution": "Work completed"}
+        )
+        fsm_order.action_complete()
+
+        # Invoice the order
+        invoice = self.sale_order_1._create_invoices()
+        # 1 invoices created
+        self.assertEqual(
+            len(invoice.ids), 1, "FSM Sale: Sale Order 1 should create 1 invoice"
+        )
+        self.assertTrue(
+            fsm_order in invoice.fsm_order_ids,
+            "FSM Sale: Invoice should be linked to FSM Order",
+        )
+
+    def test_sale_order_2(self):
+        """Test the sales order 2 flow from sale to invoice.
+        - One FSM order linked to the Sale Order Line should be created.
+        - The FSM Order should update qty_delivered when completed.
+        - One Invoice linked to the FSM Order should be created.
+        """
+        sol = self.sol_service_per_line_1
+        # Confirm the sale order
+        self.sale_order_2.action_confirm()
+        # 1 order created
+        self.assertEqual(
+            len(self.sale_order_2.fsm_order_ids.ids),
+            1,
+            "FSM Sale: Sale Order 2 should create 1 FSM Order",
+        )
+        FSM_Order = self.env["fsm.order"]
+        fsm_order = FSM_Order.search([("id", "=", sol.fsm_order_id.id)])
+        # SOL linked to FSM order
+        self.assertTrue(
+            sol.fsm_order_id.id == fsm_order.id,
+            "FSM Sale: Sale Order 2 Line not linked to FSM Order",
+        )
+
+        # Complete the FSM order
+        if self._isp_account_installed():
+            fsm_order = self._fulfill_order(fsm_order)
+        fsm_order.write(
+            {"date_end": fields.Datetime.today(), "resolution": "Work completed"}
+        )
+        fsm_order.action_complete()
+        # qty delivered should be updated
+        self.assertTrue(
+            sol.qty_delivered == sol.product_uom_qty,
+            "FSM Sale: Sale Order Line qty delivered not equal to qty ordered",
+        )
+
+        # Invoice the order
+        invoice = self.sale_order_2._create_invoices()
+        # 1 invoice created
+        self.assertEqual(
+            len(invoice.ids), 1, "FSM Sale: Sale Order 2 should create 1 invoice"
+        )
+        self.assertTrue(
+            fsm_order in invoice.fsm_order_ids,
+            "FSM Sale: Invoice should be linked to FSM Order",
+        )
+
+    def test_sale_order_3(self):
+        """Test sale order 3 flow from sale to invoice.
+        - An FSM order should be created for each Sale Order Line.
+        - The FSM Order should update qty_delivered when completed.
+        - An Invoice linked to each FSM Order should be created.
```

Comment:
I don't get this.

All fsm orders shoudn't be in the same invoice ?

Reply:
Ping @wolfhall @max3903

Reply:
Asked @osi-scampbell about this one. Some SOs will trigger creation of multiple FSOs. One (or more) of those FSOs can include labor that wasn't included on the original SO. Therefore, it creates its own invoice for those lines not included on the original SO. This is an intended feature

Reply:
Yes, it is an intended feature.  However, I don't prefer it for my usage.

Does anyone have suggestions for how we can accomplish both ways?  Maybe a setting in the configuration panel that will either split invoices or keep them as one?

Reply:
My thoughts exactly. It would be nice to add a feature that allows a choice between the existing functionality (each FSO generates an invoice) and consolidating invoices for all FSOs created from the same SO. But that's not really part of migration, it's a new feature. I'm not familiar with OCA best practices in this case, but IMO it's best to leave new features separate from the migration PR

Reply:
Yes agreed @RLeeOSI New feature will not be part of migration. To introduce new feature, create a issue, represent blueprint and submit PR.

Reply:
>One (or more) of those FSOs can include labor that wasn't included on the original SO. Therefore, it creates its own invoice for those lines not included on the original SO. This is an intended feature

I think it can be implemented with existing modules:
You can add sale order line with a product.type=service (labor) automatically when adding a product.type=fsm thanks to product-pack ( https://github.com/oca/product-pack ) or with more advanced settings ( https://github.com/akretion/sale-configurator ).
You can then track the progression of service with timesheets. It will be more aligned with other parts of odoo (stock, mrp).

Even if this proposition do not work for your use case, are you really sure this feature need to be included in fieldservice_sale and not in his own module  ?
May be I'm wrong, but from point of view this feature is too specific to be included in fieldservice_sale. It can be easily extracted in another module, and letting it in fieldservice_sale adds complexity like now during a migration.


>But that's not really part of migration, it's a new feature.
Let's say it's part of a migration cleanup/refactoring. Doing it now is may be the best moment to introduce breaking change like dependency graph and modules organization.

Reply:
@hparfr I am good with migration cleanup/refactoring. But can you provide it with migration script and unit test?